torch/csrc/dynamo/guards.cpp

```diff
@@ -0,0 +1,378 @@
+    return NULL;
+  }
+  auto& checks = *self->checks;
+  auto len = PyTuple_GET_SIZE(args);
+
+  if (static_cast<decltype(len)>(checks.size()) != len) {
+    PyErr_SetString(PyExc_TypeError, "wrong length");
+    return NULL;
+  }
+
+  LocalState state;
+
+  for (auto i : c10::irange(len)) {
+    PyObject* item = PyTuple_GET_ITEM(args, i);
+    if (Py_TYPE(item) != checks[i].pytype) {
+      Py_RETURN_FALSE;
+    }
+    if (!checks[i].check(state, THPVariable_Unpack(item))) {
+      Py_RETURN_FALSE;
+    }
+  }
+
+  Py_RETURN_TRUE;
+}
+
+PyObject* TensorGuards_check_verbose(
+    TensorGuards* self,
+    PyObject* args,
+    PyObject* kwargs) {
+  if (!PyTuple_CheckExact(args)) {
+    PyErr_SetString(PyExc_TypeError, "expected tuple()");
+    return NULL;
+  }
+  auto& checks = *self->checks;
+  auto len = PyTuple_GET_SIZE(args);
+
+  if (static_cast<decltype(len)>(checks.size()) != len) {
+    PyErr_SetString(PyExc_TypeError, "wrong length");
+    return NULL;
+  }
+
+  PyObject* tensor_check_names_py =
+      PyDict_GetItemString(kwargs, "tensor_check_names");
+  if (tensor_check_names_py == NULL) {
+    PyErr_SetString(PyExc_TypeError, "missing tensor_check_names kwarg");
+    return NULL;
+  }
+
+  if (!PyList_Check(tensor_check_names_py)) {
+    PyErr_SetString(PyExc_TypeError, "tensor_check_names kwarg must be a list");
+    return NULL;
+  }
+
+  auto names_size = PyList_Size(tensor_check_names_py);
+  if (names_size != static_cast<decltype(names_size)>(checks.size())) {
+    PyErr_SetString(
+        PyExc_TypeError,
+        "tensor_check_names should be the same size as # tensors");
+    return NULL;
+  }
+
+  std::vector<std::string> tensor_check_names;
+  tensor_check_names.reserve(names_size);
+  for (auto i : c10::irange(names_size)) {
+    PyObject* value = PyList_GetItem(tensor_check_names_py, i);
+    if (!PyUnicode_Check(value)) {
+      PyErr_SetString(
+          PyExc_TypeError, "tensor_check_names must only contain strings");
+      return NULL;
+    }
+    tensor_check_names.emplace_back(PyUnicode_AsUTF8(value));
+  }
+
+  LocalState state;
+  for (auto i : c10::irange(len)) {
+    PyObject* item = PyTuple_GET_ITEM(args, i);
+    if (Py_TYPE(item) != checks[i].pytype) {
+      std::stringstream fail_reason;
+      PyObject* type_str = PyObject_Str(PyObject_Type(item));
+      fail_reason << "expected type of '" << tensor_check_names[i]
+                  << "' to be a tensor type, ";
+      if (!type_str) {
+        fail_reason << "but found a different type";
+      } else {
+        fail_reason << "' but found " << PyUnicode_AsUTF8(type_str);
+      }
+      return Py_BuildValue("s", fail_reason.str().c_str());
+    }
+    std::string fail_reason = checks[i].check_verbose(
+        state, THPVariable_Unpack(item), tensor_check_names[i]);
+    if (fail_reason.length() > 0) {
+      return Py_BuildValue("s", fail_reason.c_str());
+    }
+  }
+
+  Py_RETURN_TRUE;
+}
+
+static PyMethodDef TensorGuards_methods[] = {
```

Comment:
nit: these bindings would be a lot simpler via pybind.

Reply:
pybind is crazy slow.

Reply:
yes :/